src/main/java/org/zalando/nakadi/filters/TracingFilter.java

```diff
@@ -0,0 +1,226 @@
+        final RequestInfo requestInfo = new RequestInfo(request, System.currentTimeMillis());
+        final RequestType requestType = getRequestType(requestInfo);
+        //Skip filter in case of non traced request
+        final Map<String, String> requestHeaders = Collections.list(request.getHeaderNames())
+                .stream()
+                .collect(Collectors.toMap(h -> h, request::getHeader));
+
+        final SpanContext spanContext = GlobalTracer.get()
+                .extract(HTTP_HEADERS, new TextMapExtractAdapter(requestHeaders));
+        final Span baseSpan;
+
+
+        if (spanContext != null) {
+            if (requestType.equals(RequestType.COMMIT)) {
+                baseSpan = TracingService.getNewSpanWithReference(requestType.getOperationName(),
+                        requestInfo.requestTime, spanContext);
+            } else {
+                baseSpan = TracingService.getNewSpanWithParent(requestType.getOperationName(),
+                        requestInfo.requestTime, spanContext);
+            }
+        } else {
+            baseSpan = TracingService.getNewSpan(requestType.getOperationName(), requestInfo.requestTime);
+        }
+
+        try {
+            final Scope scope = TracingService.activateSpan(baseSpan, false);
+            TracingService.setCustomTags(scope.span(),
+                    ImmutableMap.<String, Object>builder()
+                            .put("client_id", requestInfo.user)
+                            .put("http.url", requestInfo.path + requestInfo.query)
+                            .put("http.header.content_encoding", requestInfo.contentEncoding)
+                            .put("http.header.accept_encoding", requestInfo.acceptEncoding)
+                            .put("http.header.user_agent", requestInfo.userAgent)
+                            .build());
+            request.setAttribute("span", baseSpan);
+            //execute request
+            filterChain.doFilter(request, response);
+            if (request.isAsyncStarted()) {
+                final String flowId = FlowIdUtils.peek();
+                request.getAsyncContext().addListener(new AsyncRequestListener(request, response,
+                        requestInfo.requestTime, flowId, baseSpan, requestType));
+            }
+        } finally {
+            if (!request.isAsyncStarted()) {
+                traceRequest(requestInfo, response.getStatus(), baseSpan, requestType);
+            }
+            final Map<String, String> spanContextToInject = new HashMap<>();
+            GlobalTracer.get().inject(baseSpan.context(),
+                    HTTP_HEADERS, new TextMapInjectAdapter(spanContextToInject));
+            response.setHeader(SPAN_CONTEXT, spanContextToInject.toString());
+            baseSpan.finish();
+        }
+    }
+
+
+    private RequestType getRequestType(final RequestInfo requestLogInfo) {
+        if (requestLogInfo.path != null && "POST".equals(requestLogInfo.method) &&
```

Comment:
Maybe we can add `if(requestLogInfo.path == null){ return RequestType.OTHER; }` above and remove the null check from individual if cases.